### ISSUE TITLE
Feature Request: use try/catch statement in safe approve

### ISSUE THREAD

**Issue by gabririgo** [RigoBlock/v3-contracts#391]
Because some old tokens do not return a boolean when the approve method is called, it was necessary to develop a safe approval method that would be able to correctly set the approval and revert as desired. However, this requires using a low-level call which, while being efficient from a gas perspective, is undesirable.

Using a try/catch statement would help achieve the same result without using low-level calls. While this should be generally more gas expensive, it will allow some explicit functions/rules and potentially remove the requirement of tokenIn being a smart contract, thus removing at least 1 if statement.

The proposed implementation would work as follows:
```
function _safeApprove(address token, address spender, uint256 amount) internal {
  try IToken(token).approve(spender, amount) returns (bool success) {
    assert(success);
  } catch {
    try IToken(token).approve(spender, amount) {

    } catch { revert(); }
  }
}
```
The method will first try to approve a standard ERC20. If the call does not revert, it will assert that a boolean is returned. If the boolean is not returned, the target is either an old ERC20, does not implement approval, or is a standard wallet. The following try/catch statement will then try to execute the `approve` transaction, which will fail silently if either the target does not implement the method, or if it is not a smart contract, thus removing the need to verify the condition.

This new way of asserting approvals has the added benefit of not having to make assumptions about a swap transaction failing silently. Currently, a Uniswap swap transaction will either succeed to fail with a revert, thus not allowing potential approvals inflation. However, we have to make an assumption of how the transaction will be handled in an external protocol. Handling the case explicitly will help avoid such possibilities, or at least greatly reduce them.

**Comment by gabririgo** [comment]
notice: gas benchmarking needed

**Comment by asujan205** [comment]
@gabririgo  can i try this one?

**Comment by gabririgo** [comment]
> @gabririgo can i try this one?

sure

**Comment by gabririgo** [comment]
fixed by #608 . Further implementations of safeApprove will use this method.
